### PR TITLE
MPS: add SDPA fused backend choice for native prefill attention

### PR DESCRIPTION
**Pull request by lyonsno**
## Summary

Add an MPS dispatch for `_fused_sdp_choice` so eligible forward-inference MPS SDPA prefill rows are reported through PyTorch's existing fused SDPA backend choice.

This makes MPS SDPA backend control more truthful for the supported native prefill path: eligible rows can be selected as `SDPBackend.FLASH_ATTENTION`, ineligible Flash-only rows fail instead of silently running Math, and forced `SDPBackend.ERROR` fails closed for concrete MPS inputs.

For MPS, `SDPBackend.FLASH_ATTENTION` here means the supported native MPS prefill attention path exposed through PyTorch's existing fused backend enum. This PR does not add a new Metal FlashAttention kernel and does not claim CUDA FlashAttention feature parity, dropout/backward support, or broad training support. Forced `SDPBackend.MATH` continues to use the existing MPS math SDPA entry point, which may internally choose MPS kernels including native prefill for eligible rows.

## Changes

- Add MPS `_fused_sdp_choice` dispatch registration.
- Share a strict MPS native-prefill eligibility predicate between chooser and public MPS SDPA execution.
- Route forced/selected MPS fused attention through the existing native MPS prefill path only for supported forward-inference rows.
- Make forced Flash reject unsupported MPS rows instead of silently falling back to Math.
- Make forced Error fail closed for concrete MPS SDPA inputs.
- Preserve symbolic/export-oriented MPS SDPA behavior while respecting disabled Math / forced Error policy.
- Add regression coverage for forced-backend behavior, unsupported shape/mask/grad cases, symbolic export policy, and rank/broadcast edge cases.

## Verification

Local verification:

- `python test/test_transformers.py TestSDPAMPS.test_sdpa_export_prefill_eligible_mps_public_api_mps TestSDPAMPS.test_sdpa_export_forced_error_mps_public_api_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_prefill_flash_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_prefill_flash_matches_math_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_flash_only_rejects_prefill_ineligible_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_flash_only_rejects_prefill_shape_contract_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_default_rejects_prefill_shape_contract_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_rejects_nonbroadcastable_same_product_batch_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_rank5_singleton_batch_broadcast_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_rejects_grad_mask_prefill_mps TestSDPAFailureModesMPS.test_fused_sdp_choice_mps_flash_only_rejects_causal_with_explicit_mask_mps TestSDPAFailureModesMPS.test_dispatch_fails_no_backend_with_mps_chooser_mps -v` passed: `Ran 12 tests`, `OK`.
- `python test/test_transformers.py TestSDPAFailureModesMPS -v` passed: `Ran 52 tests`, `OK (skipped=33)`.
- `git diff --check` passed.

### CODE REVIEW

**Review by Isalia20 — CHANGES_REQUESTED** [review]
Duplicate of https://github.com/pytorch/pytorch/pull/184585

**Comment by lyonsno** [comment]
Thanks for pointing me to #184585. Agreed that this overlaps with that PR, and #184585 is the better vehicle since it covers the MPS SDPA backend dispatch surface more broadly.

I will close this one in favor of #184585. While comparing them I noticed a few edge-case coverage items around forced-backend policy, rank > 4 leading-batch broadcast, symbolic export, and grad-requiring masks; I will follow up on #184585 separately if those are still useful.
